### PR TITLE
Draft in progress: address #71 `with_nix()`

### PR DESCRIPTION
**Pull request by philipp-baumann**
@b-rodrigues not for merging yet ;-)

### CODE REVIEW

File: dev/build_envs.Rmd [hunk 1]
```diff
@@ -807,3 +818,135 @@ nix_build_exit_msg <- function(x) {
...
+  # 3) run expression in nix session using formals/input args;
+  #    mostly metaprogramming
+  
+  # 4) serialize resulted output from evaluating function in `expr`
+  
+  # 5) deserialize final output of `expr` evaluated in nix-shell
+  #    into host R session
+  
+  cmd <- switch(program,
+    # do 2), 3), 4) in nix-shell-R session (check how to deal with shellHook)
```
Comment (b-rodrigues) [inline dev/build_envs.Rmd:895]: I guess we should run the R expression with R --vanilla?

**Comment by philipp-baumann** [comment]
The logic is now even simpler. Here a short example of running

```{r}
with_nix(expr = function(m = mtcars) nrow(mtcars), exec_mode = "non-blocking")
```
In step 2 (get formals/arguments in nix-R), we now just write an R script file. This has two advantages. 1) we can save us some eval-parse round, and 2) it is way easier to run multi-line R code via `Rscript`, in fact it is recommended to use `source()` in those cases.

Here's the resulting scriptfile in the `temp_folder`:

```{sh}
philipp@philipp-private:/tmp/RtmpOPLmZu$ cat with_nix_r.R 
#!/bin/sh
temp_dir <- "/tmp/RtmpOPLmZu"
r_version_num <- paste0(R.version$major, ".", R.version$minor)
# assign `args_vec` as in c(...) form.
args_vec <- c(m = "mtcars")
# assign elements of `args_vec` individually
m <- "mtcars"
# actual deserialization step
read_args <- function (args_vec, temp_dir) 
{
    for (i in seq_len(length(args_vec))) {
        nm <- args_vec[i]
        obj <- names(args_vec)[i]
        assign(nm, readRDS(file = file.path(temp_dir, paste0(obj, ".Rds"))))
        cat(paste0("reading ", obj, ".Rds"))
    }
}
# evaluate function
read_args(args_vec, temp_dir)
ls()
```

**Comment by philipp-baumann** [comment]
with_nix_part1

**Comment by philipp-baumann** [comment]
Finally one step further :-) We can now evaluate first part of R expressions via `Rscript` using `nix-shell`, which is invoked via `sys::exec_*()`.

Here is the current output serializing.

```{r}
> with_nix(
+   expr = function(m = mtcars) nrow(mtcars),
+   exec_mode = "non-blocking",
+   project_path = "inst/extdata/with_nix"
+ )
==> Running deparsed expression via `nix-shell` in non-blocking mode:
# -----------------------------------------------------------------------------
temp_dir <- "/tmp/Rtmpc0gLAe"
r_version_num <- paste0(R.version$major, ".", R.version$minor)
# assign `args_vec` as in c(...) form.
args_vec <- c(m = "mtcars")
# assign elements of `args_vec` individually
m <- "mtcars"
# actual deserialization step
read_args <- function (args_vec, temp_dir) 
{
    for (i in seq_len(length(args_vec))) {
        nm <- args_vec[i]
        obj <- names(args_vec)[i]
        assign(nm, readRDS(file = file.path(temp_dir, paste0(obj, ".Rds"))))
        cat(paste0("reading ", obj, ".Rds"))
    }
}
# evaluate function
read_args(args_vec, temp_dir)
cat("
* using Nix with R version", paste0(R.version$major, ".", R.version$minor), "
")
cat("* writing R script evaluated via `Rscript` in `nix-shell`:", "/tmp/Rtmpc0gLAe/with_nix_r.R")
cat("
`sessionInfo()` output:
")
capture.output(sessionInfo())
# ------------------------------------------------------------------------------


==> Process ID (PID) is 22655.
==> Receiving stdout and stderr streams...
reading m.Rds
* using Nix with R version 4.3.1 
* writing R script evaluated via `Rscript` in `nix-shell`: /tmp/Rtmpc0gLAe/with_nix_r.R
`sessionInfo()` output:
 [1] "R version 4.3.1 (2023-06-16)"                                                                           
 [2] "Platform: x86_64-pc-linux-gnu (64-bit)"                                                                 
 [3] "Running under: Ubuntu 22.04.3 LTS"                                                                      
 [4] ""                                                                                                       
 [5] "Matrix products: default"                                                                               
 [6] "BLAS/LAPACK: /nix/store/8lqs4s515p2597vl838mrzb03qgwwvag-blas-3/lib/libblas.so.3;  LAPACK version 3.9.0"
 [7] ""                                                                                                       
 [8] "locale:"                                                                                                
 [9] " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              "                                             
[10] " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    "                                             
[11] " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   "                                             
[12] " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 "                                             
[13] " [9] LC_ADDRESS=C               LC_TELEPHONE=C            "                                             
[14] "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       "                                             
[15] ""                                                                                                       
[16] "time zone: Europe/Zurich"                                                                               
[17] "tzcode source: system (glibc)"                                                                          
[18] ""                                                                                                       
[19] "attached base packages:"                                                                                
[20] "[1] stats     graphics  grDevices utils     datasets  methods   base     "                              
[21] ""                                                                                                       
[22] "loaded via a namespace (and not attached):"                                                             
[23] "[1] compiler_4.3.1"                                                                                     

==> `expr` succeeded!
```

**Comment by philipp-baumann** [comment]
We can now also handle args of function in  `expr` that are any object (language). Currently, `expr` is executed in Nix env with formals that are read from disk (saved previously in host R session).

```{r}
> with_nix(
+   expr = function(m = data.frame(a = "bla")) nrow(m),
+   exec_mode = "non-blocking",
+   project_path = "inst/extdata/with_nix"
+ )
==> Running deparsed expression via `nix-shell` in non-blocking mode:

# -----------------------------------------------------------------------------
cat("
* wrote R script evaluated via `Rscript` in `nix-shell`:", "/tmp/RtmpK2rpdF/with_nix_r.R")
temp_dir <- "/tmp/RtmpK2rpdF"
r_version_num <- paste0(R.version$major, ".", R.version$minor)
cat("
* using Nix with R version", r_version_num, "
")
# assign `args_vec` as in c(...) form.
args_vec <- c("m")
# deserialize arguments from disk
read_args <- function (args_vec, temp_dir) 
{
    for (i in seq_len(length(args_vec))) {
        nm <- args_vec[i]
        obj <- args_vec[i]
        assign(x = nm, value = readRDS(file = file.path(temp_dir, paste0(obj, ".Rds"))), envir = .GlobalEnv)
        cat(paste0("* reading ", obj, ".Rds"))
    }
}
# evaluate function
read_args(args_vec, temp_dir)
# deparse and run function given in `expr` arg
run_expr <- function (m = data.frame(a = "bla")) 
nrow(m)

# execute function call in `expr` with list of correct args
lst <- as.list(args_vec)
names(lst) <- args_vec
lst <- lapply(lst, as.name)
nix_out <- do.call(run_expr, lst)
cat(paste("
* called `expr` with args", args_vec)) 
cat("
* the following objects are in the global environment:
")
cat(ls())
cat("
* `sessionInfo()` output:
")
capture.output(sessionInfo())
# ------------------------------------------------------------------------------


==> Process ID (PID) is 204706.
==> Receiving stdout and stderr streams...

* wrote R script evaluated via `Rscript` in `nix-shell`: /tmp/RtmpK2rpdF/with_nix_r.R
* using Nix with R version 4.3.1 
* reading m.Rds
* called `expr` with args m
* the following objects are in the global environment:
args_vec lst m nix_out r_version_num read_args run_expr temp_dir
* `sessionInfo()` output:
 [1] "R version 4.3.1 (2023-06-16)"                                                                           
 [2] "Platform: x86_64-pc-linux-gnu (64-bit)"                                                                 
 [3] "Running under: Ubuntu 22.04.3 LTS"                                                                      
 [4] ""                                                                                                       
 [5] "Matrix products: default"                                                                               
 [6] "BLAS/LAPACK: /nix/store/8lqs4s515p2597vl838mrzb03qgwwvag-blas-3/lib/libblas.so.3;  LAPACK version 3.9.0"
 [7] ""                                                                                                       
 [8] "locale:"                                                                                                
 [9] " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              "                                             
[10] " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    "                                             
[11] " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   "                                             
[12] " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 "                                             
[13] " [9] LC_ADDRESS=C               LC_TELEPHONE=C            "                                             
[14] "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       "                                             
[15] ""                                                                                                       
[16] "time zone: Europe/Zurich"                                                                               
[17] "tzcode source: system (glibc)"                                                                          
[18] ""                                                                                                       
[19] "attached base packages:"                                                                                
[20] "[1] stats     graphics  grDevices utils     datasets  methods   base     "                              
[21] ""                                                                                                       
[22] "loaded via a namespace (and not attached):"                                                             
[23] "[1] compiler_4.3.1"                                                                                     

==> `expr` succeeded!
```

**Comment by philipp-baumann** [comment]
Current commit works when no extra packages are loaded. With custom packages, a major point of using `with_nix()` is to test software environments, their reproducibility and potential breaking changes. Hence, we need static code analysis further export and globals. 

- If globals define functions and are in a packages environment, those packages need to be exported. If globals are functions defined in the global environment, then those functions need to be deparsed added to the nix R script.
- If globals of `expr` are defined outside of the function environment, then `with_nix()` should stop with an informal message. This is a design choice to force well-defined scoping of variables when calling `expr`.

**Comment by philipp-baumann** [comment]
simple case (works)

```{r}
> with_nix(
+   expr = function(m = data.frame(a = "bla")) nrow(m),
+   exec_mode = "non-blocking",
+   project_path = "inst/extdata/with_nix"
+ )
* checking code in `expr` for potential problems:
 `codetools::checkUsage(fun = expr)`

==> Running deparsed expression via `nix-shell` in non-blocking mode:

# -----------------------------------------------------------------------------
cat("
* wrote R script evaluated via `Rscript` in `nix-shell`:", "/tmp/Rtmpflg1Dg/with_nix_r.R")
temp_dir <- "/tmp/Rtmpflg1Dg"
r_version_num <- paste0(R.version$major, ".", R.version$minor)
cat("
* using Nix with R version", r_version_num, "
")
# assign `args_vec` as in c(...) form.
args_vec <- c("m")
# deserialize arguments from disk
for (i in seq_len(length(args_vec))) {
    nm <- args_vec[i]
    obj <- args_vec[i]
    assign(x = nm, value = readRDS(file = file.path(temp_dir, paste0(obj, ".Rds"))))
    cat(paste0("* reading ", obj, ".Rds"))
}
# deparse and run function given in `expr` arg
run_expr <- function (m = data.frame(a = "bla")) 
nrow(m)

# execute function call in `expr` with list of correct args
lst <- as.list(args_vec)
names(lst) <- args_vec
lst <- lapply(lst, as.name)
nix_out <- do.call(run_expr, lst)
cat(paste("
* called `expr` with args", args_vec)) 
cat("
* the following objects are in the global environment:
")
cat(ls())
cat("
* `sessionInfo()` output:
")
capture.output(sessionInfo())
# ------------------------------------------------------------------------------


==> Process ID (PID) is 17418.
==> Receiving stdout and stderr streams...

* wrote R script evaluated via `Rscript` in `nix-shell`: /tmp/Rtmpflg1Dg/with_nix_r.R
* using Nix with R version 4.3.1 
* reading m.Rds
* called `expr` with args m
* the following objects are in the global environment:
args_vec i lst m nix_out nm obj r_version_num run_expr temp_dir
* `sessionInfo()` output:
 [1] "R version 4.3.1 (2023-06-16)"                                                                           
 [2] "Platform: x86_64-pc-linux-gnu (64-bit)"                                                                 
 [3] "Running under: Ubuntu 22.04.3 LTS"                                                                      
 [4] ""                                                                                                       
 [5] "Matrix products: default"                                                                               
 [6] "BLAS/LAPACK: /nix/store/8lqs4s515p2597vl838mrzb03qgwwvag-blas-3/lib/libblas.so.3;  LAPACK version 3.9.0"
 [7] ""                                                                                                       
 [8] "locale:"                                                                                                
 [9] " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              "                                             
[10] " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    "                                             
[11] " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   "                                             
[12] " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 "                                             
[13] " [9] LC_ADDRESS=C               LC_TELEPHONE=C            "                                             
[14] "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       "                                             
[15] ""                                                                                                       
[16] "time zone: Europe/Zurich"                                                                               
[17] "tzcode source: system (glibc)"                                                                          
[18] ""                                                                                                       
[19] "attached base packages:"                                                                                
[20] "[1] stats     graphics  grDevices utils     datasets  methods   base     "                              
[21] ""                                                                                                       
[22] "loaded via a namespace (and not attached):"                                                             
[23] "[1] compiler_4.3.1"                                                                                     

==> `expr` succeeded!
```

**Comment by philipp-baumann** [comment]
This is an easy example that is a bit artificial but shows what is needed in terms of code functionality so that is correctly runs in nix R:

```{r}
with_nix(
  expr = function(m = data.frame(a = "bla")) nrow(m),
  exec_mode = "non-blocking",
  project_path = "inst/extdata/with_nix"
)

nrow_mtcars <- function(m = mtcars) {
  a <- datasets::mtcars |> dplyr::filter(cyl == 6L)
  nrow_impl(a)
}

nrow_impl <- function(a) nrow(a)
```

**Comment by philipp-baumann** [comment]
[[f23de2d](https://github.com/b-rodrigues/rix/pull/76/commits/f23de2d16a88093595ab64eab9ef7d8c4525adcc)]https://github.com/b-rodrigues/rix/pull/76/commits/f23de2d16a88093595ab64eab9ef7d8c4525adcc) is a first step towards boilderplating code that loads extra packages that are defined in globals found in function given by `expr` later in the Nix-R session. 

This means we can soon deal with things like:

```r
library("dplyr")

nrow_mtcars <- function(m = mtcars) {
  a <- datasets::mtcars |> filter(cyl == 6L)
  nrow_impl(a)
}

nrow_impl <- function(a) nrow(a)
```

**Comment by philipp-baumann** [comment]
@b-rodrigues FYI, I am now testing this example. If you have further scenarios of a typical `expr` setting in terms of the abstract syntax tree, the environment stacking etc., would be great to hear :-)

```r
library("dplyr")

env_a <- new.env()
env_a$key <- "value"

nrow_impl <- function(a) {
  not_returned <- data.table(a = 1:3)
  cat("`not_returned`", "has", nrow(not_used), "rows.")
  nrow(a)
}

nrow_mtcars <- function(m = mtcars) {
  a <- datasets::mtcars |> filter(cyl == 6L)
  cat("Object outside of function scope is of class: ", class(env_a))
  nrow_impl(a)
}

with_nix(
  expr = nrow_mtcars,
  exec_mode = "non-blocking",
  project_path = "inst/extdata/with_nix"
)
```

**Comment by philipp-baumann** [comment]
[ ] make R-nix code layer bullet proof for this case:

```r
```

**Comment by philipp-baumann** [comment]
On macOS, `nix-shell` is not found, because path seems not exported. 

```r
sys::exec_internal(cmd = "/nix/var/nix/profiles/default/bin/nix-shell", cmd_rnix_deparsed)
```

This is the session info:

```r
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       macOS Sonoma 14.0
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Zurich
 date     2023-11-15
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────────────
 package     * version date (UTC) lib source
 attachment    0.4.0   2023-05-31 [1] CRAN (R 4.3.0)
 backports     1.4.1   2021-12-13 [1] CRAN (R 4.3.0)
 brio          1.1.3   2021-11-30 [1] CRAN (R 4.3.0)
 cachem        1.0.8   2023-05-01 [1] CRAN (R 4.3.0)
 callr         3.7.3   2022-11-02 [1] CRAN (R 4.3.0)
 checkmate     2.2.0   2023-04-27 [1] CRAN (R 4.3.0)
 cli           3.6.1   2023-03-23 [1] CRAN (R 4.3.0)
 codetools     0.2-19  2023-02-01 [1] CRAN (R 4.3.1)
 crayon        1.5.2   2022-09-29 [1] CRAN (R 4.3.0)
 desc          1.4.2   2022-09-08 [1] CRAN (R 4.3.0)
 devtools      2.4.5   2022-10-11 [1] CRAN (R 4.3.0)
 digest        0.6.33  2023-07-07 [1] CRAN (R 4.3.0)
 dplyr         1.1.3   2023-09-03 [1] CRAN (R 4.3.0)
 ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.3.0)
 evaluate      0.22    2023-09-29 [1] CRAN (R 4.3.1)
 fansi         1.0.5   2023-10-08 [1] CRAN (R 4.3.1)
 fastmap       1.1.1   2023-02-24 [1] CRAN (R 4.3.0)
 fs            1.6.3   2023-07-20 [1] CRAN (R 4.3.0)
 fusen         0.5.2   2023-08-17 [1] CRAN (R 4.3.0)
 generics      0.1.3   2022-07-05 [1] CRAN (R 4.3.0)
 glue          1.6.2   2022-02-24 [1] CRAN (R 4.3.0)
 hms           1.1.3   2023-03-21 [1] CRAN (R 4.3.0)
 htmltools     0.5.6.1 2023-10-06 [1] CRAN (R 4.3.1)
 htmlwidgets   1.6.2   2023-03-17 [1] CRAN (R 4.3.0)
 httpuv        1.6.11  2023-05-11 [1] CRAN (R 4.3.0)
 httr          1.4.7   2023-08-15 [1] CRAN (R 4.3.0)
 jsonlite      1.8.7   2023-06-29 [1] CRAN (R 4.3.0)
 knitr         1.44    2023-09-11 [1] CRAN (R 4.3.0)
 later         1.3.1   2023-05-02 [1] CRAN (R 4.3.0)
 lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.3.0)
 magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.3.0)
 memoise       2.0.1   2021-11-26 [1] CRAN (R 4.3.0)
 mime          0.12    2021-09-28 [1] CRAN (R 4.3.0)
 miniUI        0.1.1.1 2018-05-18 [1] CRAN (R 4.3.0)
 parsermd      0.1.2   2021-05-20 [1] CRAN (R 4.3.0)
 pillar        1.9.0   2023-03-22 [1] CRAN (R 4.3.0)
 pkgbuild      1.4.2   2023-06-26 [1] CRAN (R 4.3.0)
 pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.3.0)
 pkgload       1.3.3   2023-09-22 [1] CRAN (R 4.3.1)
 prettyunits   1.2.0   2023-09-24 [1] CRAN (R 4.3.1)
 processx      3.8.2   2023-06-30 [1] CRAN (R 4.3.0)
 profvis       0.3.8   2023-05-02 [1] CRAN (R 4.3.0)
 promises      1.2.1   2023-08-10 [1] CRAN (R 4.3.0)
 ps            1.7.5   2023-04-18 [1] CRAN (R 4.3.0)
 purrr         1.0.2   2023-08-10 [1] CRAN (R 4.3.0)
 R6            2.5.1   2021-08-19 [1] CRAN (R 4.3.0)
 Rcpp          1.0.11  2023-07-06 [1] CRAN (R 4.3.0)
 readr         2.1.4   2023-02-10 [1] CRAN (R 4.3.0)
 remotes       2.4.2.1 2023-07-18 [1] CRAN (R 4.3.0)
 rlang         1.1.1   2023-04-28 [1] CRAN (R 4.3.0)
 rmarkdown     2.25    2023-09-18 [1] CRAN (R 4.3.1)
 roxygen2      7.2.3   2022-12-08 [1] CRAN (R 4.3.0)
 rprojroot     2.0.3   2022-04-02 [1] CRAN (R 4.3.0)
 rstudioapi    0.15.0  2023-07-07 [1] CRAN (R 4.3.0)
 sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.3.0)
 shiny         1.7.5.1 2023-10-14 [1] CRAN (R 4.3.1)
 stringi       1.7.12  2023-01-11 [1] CRAN (R 4.3.0)
 stringr       1.5.0   2022-12-02 [1] CRAN (R 4.3.0)
 sys         * 3.4.2   2023-05-23 [1] CRAN (R 4.3.0)
 testthat      3.2.0   2023-10-06 [1] CRAN (R 4.3.1)
 tibble        3.2.1   2023-03-20 [1] CRAN (R 4.3.0)
 tidyr         1.3.0   2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect    1.2.0   2022-10-10 [1] CRAN (R 4.3.0)
 tzdb          0.4.0   2023-05-12 [1] CRAN (R 4.3.0)
 urlchecker    1.0.1   2021-11-30 [1] CRAN (R 4.3.0)
 usethis       2.2.2   2023-07-06 [1] CRAN (R 4.3.0)
 utf8          1.2.3   2023-01-31 [1] CRAN (R 4.3.0)
 vctrs         0.6.4   2023-10-12 [1] CRAN (R 4.3.1)
 withr         2.5.1   2023-09-26 [1] CRAN (R 4.3.1)
 xfun          0.40    2023-08-09 [1] CRAN (R 4.3.0)
 xml2          1.3.5   2023-07-06 [1] CRAN (R 4.3.0)
 xtable        1.8-4   2019-04-21 [1] CRAN (R 4.3.0)
 yaml          2.3.7   2023-01-23 [1] CRAN (R 4.3.0)

 [1] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library
```

**Comment by philipp-baumann** [comment]
Testing current `with_nix()` via

```r
library("dplyr")
library("data.table")

env_a <- new.env()
env_a$key <- "value"

nrow_impl <- function(a) {
  not_returned <- data.table(a = 1:3)
  cat("`not_returned`", "has", nrow(not_used), "rows.")
  df_check(a)
  nrow(a)
}

df_check <- function(x) stopifnot("x must be data.frame" = is_df(x))

# critical, not assigned
is_df <- function(x) is.data.frame(x)

nrow_mtcars <- function(m = mtcars) {
  a <- m |> filter(cyl == 6L)
  cat("Object outside of function scope is of class: ", class(env_a))
  nrow_impl(a)
  test_df(a)
}

test_df <- function(a) is_df_a(a)

is_df_a <- function(a) is.data.frame(a)

with_nix(
  expr = nrow_mtcars,
  exec_mode = "non-blocking",
  project_path = "."
)
```

**Pull request merged**